test/integration/oauth_serviceaccount_client_events_test.go

```diff
@@ -0,0 +1,437 @@
+	}
+	clientTransport, err := restclient.TransportFor(clusterAdminClientConfig)
+	testTransport := &basicAuthTransport{rt: clientTransport}
+	oauthRuntimeClient.Transport = testTransport
+
+	authorizeRequest := oauthRuntimeClient.NewAuthorizeRequest(osincli.CODE)
+	req, err := http.NewRequest("GET", authorizeRequest.GetAuthorizeUrlWithParams("").String(), nil)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	// Set up the HTTP redirect handler
+	operations := []string{}
+	jar, _ := cookiejar.New(nil)
+	directHTTPClient := &http.Client{
+		Transport: testTransport,
+		CheckRedirect: func(redirectReq *http.Request, via []*http.Request) error {
+			t.Logf("302 Location: %s", redirectReq.URL.String())
+			req = redirectReq
+			operations = append(operations, "redirect to "+redirectReq.URL.Path)
+			return nil
+		},
+		Jar: jar,
+	}
+
+	for {
+		t.Logf("%s %s", req.Method, req.URL.String())
+		operations = append(operations, req.Method+" "+req.URL.Path)
+
+		// Always set the csrf header
+		req.Header.Set("X-CSRF-Token", "1")
+		resp, err := directHTTPClient.Do(req)
+		if err != nil {
+			t.Errorf("Error %v\n%#v\n%#v", operations, jar, err)
+			return
+		}
+		defer resp.Body.Close()
+
+		// Save the current URL for reference
+		currentURL := req.URL
+
+		if resp.StatusCode == 401 {
+			// Set up a username and password once we're challenged
+			testTransport.username = adminUser
+			testTransport.password = "any-pass"
+			operations = append(operations, "received challenge")
+			continue
+		}
+
+		if expectBadRequest && resp.StatusCode == 400 {
+			responseDump, _ := httputil.DumpResponse(resp, true)
+			t.Logf("Bad Request: %s", string(responseDump))
+			return
+		}
+
+		if resp.StatusCode != 200 {
+			responseDump, _ := httputil.DumpResponse(resp, true)
+			t.Errorf("Unexpected response %s", string(responseDump))
+			return
+		}
+
+		doc, err := html.Parse(resp.Body)
+		if err != nil {
+			t.Error(err)
+			return
+		}
+		forms := htmlutil.GetElementsByTagName(doc, "form")
+		// if there's a single form, submit it
+		if len(forms) > 1 {
+			t.Errorf("More than one form encountered: %d", len(forms))
+			return
+		}
+		if len(forms) == 0 {
+			break
+		}
+		req, err = htmlutil.NewRequestFromForm(forms[0], currentURL, nil)
+		if err != nil {
+			t.Error(err)
+			return
+		}
+		operations = append(operations, "form")
+	}
+
+	select {
+	case <-authorizationCodes:
+		operations = append(operations, "code")
+	case authorizationError := <-authorizationErrors:
+		operations = append(operations, "error:"+authorizationError)
+	case <-time.After(5 * time.Second):
+		t.Error("didn't get a code or an error")
+	}
+
+	if !reflect.DeepEqual(operations, expectOperations) {
+		t.Errorf("Expected:\n%#v\nGot\n%#v", expectOperations, operations)
+	}
+
+}
+
+// TestOAuthServiceAccountClientEvent verifies that certain warning events are created when an SA is incorrectly configured
+// for OAuth
+func TestOAuthServiceAccountClientEvent(t *testing.T) {
```

Comment:
Consider moving this method to the top. IMHO it's easier to read when the test cases are listed first and test helpers go last.